Importer for Flighty flight information

jphastings opened the pull request:
A first pass at an importer for [Flighty](https://flighty.com) flight tracker.
    
A flight is modelled as:
- A `Collection`, with:
    - markdown content that describes the flight "Flight from _London Heathrow Airport_ to _Kempegowda International Airport_"
    - A timestamp equal to the takeoff time (actual, if present, scheduled if not)
    - A timespan equal to the landing time (actual, if present, scheduled if not)
    - Metadata describing the seat & cabin you sat in, the aircraft type, the flight number & air carrier
- A `Location` item in that collection for take off airport (with the appropriate timestamp)
- A `Location` item in that collection for landing airport (the 'To' airport, or the 'Diverted To', if present)
- A `Message` item with the note made for the flight, if present in the export.

Is this suitable modelling @mholt?

I've both vendored the airport database _and_ configured `go:generate` to download the latest version, if needed. Do shout if that doesn't work for you!

Finally, the UI doesn't render _very long_ distances between poins as great arcs (my guess is it renders them as if they were cartesian coordinates).

<img width="1107" alt="Timelinize rendering two flights from London to Vancouver, via Iceland — with clearly non-great arc lines" src="https://github.com/user-attachments/assets/bcb9b2ae-1115-4bcd-a3bb-f867c70d7c72" />

Let me know what you think! If you don't already have Flighty & a few flights in there, I'd recommend getting it, it's a great app 😊 (though perhaps not the paid version unless you do _a lot_ of flying!)

mholt reviewed (commented):
Need to do a more thorough review later, but, at a glance this seems pretty good. I will think on the model and keep you posted!

jphastings commented on datasources/flighty/types.go:
Ahh, I can probably unexport most of them! I left them as exported as I'd meant to ask about whether you'd be interested in me putting the airport code into an `/internal/airports` directory so other (hypotehtical) importers could use it — do you have any inclination for that?

mholt replied:
Yeah, an internal package is a neat idea!

jphastings commented:
I've rebased on top of #92 (so I can fix these pesky lint issues without having to figure out how to parallel-install an old version of `golangci-lint` 😅), removed the `Retrieval` keys, and extracted the airports DB into an `internal` package — whenever you've a spare moment I'd appreciate a review, but I know you're busy with the rest of life at the moment, and I'm away over the long weekend so there's no rush at all!

mholt commented:
Okay cool this looks better! Though it's a bit messy to read. Can we make the diff not show the changes from #92? Then I think we can merge this in.

mholt commented:
Merged #92! :+1:

jphastings commented:
All set @mholt — rebased on `main`, and ready for your review 😊

mholt commented:
Looking awesome! Will review more shortly. I'm close to having better support for place entities.

jphastings commented:
No rush! Thanks for the update 😊

jphastings commented:
Thanks @mholt! Now up to date with `main` 😊

Do you have any pointers for how I should reference the airports as entities? (I'll see what I can figure from the #101, but I'd love to make sure I'm doing it right!)

mholt commented:
Oh right, sorry. Here's what I do elsewhere:

```go
g.ToEntity(timeline.RelVisit, &timeline.Entity{
	Type: timeline.EntityPlace,
	Name: pt.Name,
	Attributes: []timeline.Attribute{
		{
			Name:      "coordinate",
			Latitude:  item.Location.Latitude,
			Longitude: item.Location.Longitude,
			Identity:  true, // TODO: Or "Identifying: true"?
			Metadata: timeline.Metadata{
				"URL": pt.Link.Href,
			},
		},
	},
})
```

In this case I have a Graph, `g`, which is an item, which I'm associating with a place, which suggests a visit to that place.

You could also just send an entity Graph instead, if there's no "visit" to that place necessarily.

I _think_ `Identity: true` is necessary, but I feel like only `Identifying: true` should be needed (although then I get `_entity` attributes, which aren't intended -- not sure if bug or that's just how it works, still investigating).

Does that make sense?

> I think I may need to swap [these takeoff and landing points](https://github.com/jphastings/timelinize/blob/b7df153b56d7d11d34b01d0c9ed8e646758b5a5d/datasources/flighty/flighty.go#L157-L175) (currently timeline.ClassLocation) to being some kind of location entity? (It can't be a timelinize.Entity, as that's for humans, particularly owners, I think)

Entities can be humans, creatures, places, or other things (organizations, offices/titles, etc)!

jphastings commented:
Oh nice; so I should keep my `ClassLocation` items to note the start and end of the journey being made, but _also_ have visits (for the fact that you've visited the airport/gate etc)

I think I'm nearly there with this 😊 The tests are tricky to write in a way that isn't _huuugely_ verbose, but also give _useful_ information on failure.

mholt commented:
Sounds about right, yep! Nice work :)

jphastings commented:
@mholt I've just added three commits that _should_ add Place Visits! I think this PR is ready for merge! Do review at your leisure 😊 

I've included the "Exact name" for the place (as an attribute), which includes terminal & gate details as well as airport name, if present. I _haven't_ included terminal & gate info in the (default) name because the lat/long used is always for the airport (as a whole), rather than the terminal/gate.

mholt commented:
Off-topic, but since you brought it up, here's some sample code from ChatGPT (sigh) to draw a great arc using turf.js:

```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <title>Great Arc on Mapbox Globe</title>
  <meta name="viewport" content="initial-scale=1,maximum-scale=1,user-scalable=no" />
  <script src="https://api.mapbox.com/mapbox-gl-js/v3.2.0/mapbox-gl.js"></script>
  <link href="https://api.mapbox.com/mapbox-gl-js/v3.2.0/mapbox-gl.css" rel="stylesheet" />
  <script src="https://cdn.jsdelivr.net/npm/@turf/turf@6/turf.min.js"></script>
  <style>
    body { margin: 0; padding: 0; }
    #map { position: absolute; top: 0; bottom: 0; width: 100%; }
  </style>
</head>
<body>
<div id="map"></div>
<script>
  mapboxgl.accessToken = 'YOUR_MAPBOX_ACCESS_TOKEN'; // Replace this

  const map = new mapboxgl.Map({
    container: 'map',
    style: 'mapbox://styles/mapbox/satellite-streets-v12',
    projection: 'globe', // key for 3D globe
    zoom: 1.5,
    center: [0, 20]
  });

  map.on('load', () => {
    // Use Turf.js to generate a great arc line
    const from = [-122.4194, 37.7749]; // San Francisco
    const to = [139.6917, 35.6895];    // Tokyo

    const arc = turf.greatCircle(from, to, {
      npoints: 100,
      properties: { name: "SF to Tokyo" }
    });

    map.addSource('great-arc', {
      type: 'geojson',
      data: arc
    });

    map.addLayer({
      id: 'great-arc-line',
      type: 'line',
      source: 'great-arc',
      layout: {
        'line-cap': 'round',
        'line-join': 'round'
      },
      paint: {
        'line-color': '#FF5733',
        'line-width': 2
      }
    });

    // Optional: atmosphere for better 3D look
    map.setFog({});
  });
</script>
</body>
</html>
```

mholt reviewed (approved):
LGTM. Nice work and thank you!

jphastings commented:
My pleasure! I'll see if there are any other interesting Datasources I might implement with what I've learned!
